cmd/serve/s3/backend.go

```diff
@@ -0,0 +1,577 @@
+// Package s3 implements a fake s3 server for rclone
+package s3
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"log"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/Mikubill/gofakes3"
+	"github.com/ncw/swift/v2"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/walk"
+	"github.com/rclone/rclone/lib/encoder"
+	"github.com/rclone/rclone/vfs"
+)
+
+var (
+	charEncoder    = encoder.EncodeCtl | encoder.EncodeLeftCrLfHtVt | encoder.EncodeRightCrLfHtVt | encoder.EncodeDel
+	emptyPrefix    = &gofakes3.Prefix{}
+	timeFormat     = "Mon, 2 Jan 2006 15:04:05.999999999 GMT"
+	tmpMetaStorage = new(sync.Map)
+)
+
+type s3Backend struct {
+	opt  *Options
+	lock sync.Mutex
+	fs   *vfs.VFS
+}
+
+// newBackend creates a new SimpleBucketBackend.
+func newBackend(fs *vfs.VFS, opt *Options) gofakes3.Backend {
```

Comment:
It should be possible to implement this without using the VFS, using the lower level fs commands only.

However the VFS provides caching etc which could be useful. It will introduce inefficiencies though of course... For example you can run ListR directly on bucket backed backends, but via the VFS it will list each directory one by one.

Reply:
Considering the high request rate of S3, it should be better to use VFS on some backends with strict rate limits. I'd added some logic for the BucketBased backend to execute direct `fs.ListR`, hope this will work.

Reply:
Is it possible to also have a VFS passthrough mode or disable it entirely? I may need it because I will load balance the S3 server using multiple nodes (they restrict it by IP traffic) so having a VFS may cause cache invalidation problem.